File: intrinsic/motion_planning/motion_planner_client.cc
// Copyright 2023 Intrinsic Innovation LLC

#include "intrinsic/motion_planning/motion_planner_client.h"

#include <cstdint>
#include <memory>
#include <optional>
#include <string>
#include <utility>
#include <vector>

#include "absl/status/statusor.h"
#include "absl/strings/string_view.h"
#include "google/protobuf/duration.pb.h"
#include "google/protobuf/empty.pb.h"
#include "grpcpp/client_context.h"
#include "intrinsic/eigenmath/types.h"
#include "intrinsic/icon/proto/joint_space.pb.h"
#include "intrinsic/logging/proto/context.pb.h"
#include "intrinsic/math/pose3.h"
#include "intrinsic/math/proto_conversion.h"
#include "intrinsic/motion_planning/conversions.h"
#include "intrinsic/motion_planning/proto/motion_target.pb.h"
#include "intrinsic/motion_planning/proto/v1/motion_planner_config.pb.h"
#include "intrinsic/motion_planning/proto/v1/motion_planner_service.grpc.pb.h"
#include "intrinsic/motion_planning/proto/v1/motion_planner_service.pb.h"
#include "intrinsic/util/eigen.h"
#include "intrinsic/util/status/status_conversion_grpc.h"
#include "intrinsic/util/status/status_macros.h"
#include "intrinsic/world/objects/kinematic_object.h"
#include "intrinsic/world/objects/transform_node.h"
#include "intrinsic/world/proto/object_world_refs.pb.h"

namespace intrinsic {
namespace motion_planning {

const MotionPlannerClient::MotionPlanningOptions&
MotionPlannerClient::MotionPlanningOptions::Defaults() {
  static const auto* defaults = new MotionPlannerClient::MotionPlanningOptions({
      .path_planning_time_out = 30,
      .path_planning_step_size = std::nullopt,
      .compute_swept_volume = false,
      .lock_motion_configuration = std::nullopt,
      .skip_fuzzy_cache_check = false,
  });

  return *defaults;
}

MotionPlannerClient::MotionPlannerClient(
    absl::string_view world_id,
    std::shared_ptr<intrinsic_proto::motion_planning::v1::MotionPlannerService::
                        StubInterface>
        motion_planner_service)
    : world_id_(world_id),
      motion_planner_service_(std::move(motion_planner_service)) {}

absl::StatusOr<MotionPlannerClient::PlanTrajectoryResult>
MotionPlannerClient::PlanTrajectory(
    const intrinsic_proto::motion_planning::v1::RobotSpecification&
        robot_specification,
    const intrinsic_proto::motion_planning::v1::MotionSpecification&
        motion_specification,
    const MotionPlanningOptions& options, const std::string& caller_id,
    const intrinsic_proto::data_logger::Context& context) {
  intrinsic_proto::motion_planning::v1::MotionPlanningRequest request;
  *request.mutable_robot_specification() = robot_specification;
  *request.mutable_motion_specification() = motion_specification;
  request.set_world_id(world_id_);
  request.set_compute_swept_volume(options.compute_swept_volume);
  request.mutable_motion_planner_config()->mutable_timeout_sec()->set_seconds(
      options.path_planning_time_out);
  const int64_t s = request.motion_planner_config().timeout_sec().seconds();
  request.mutable_motion_planner_config()->mutable_timeout_sec()->set_nanos(
      (options.path_planning_time_out - s) * 1e9);
  if (options.path_planning_step_size.has_value()) {
    request.mutable_motion_planner_config()->set_path_planning_step_size(
        *options.path_planning_step_size);
  }
  if (options.lock_motion_configuration.has_value()) {
    *request.mutable_motion_planner_config()
         ->mutable_lock_motion_configuration() =
        options.lock_motion_configuration.value();
  }
  request.set_caller_id(caller_id);
  *request.mutable_context() = context;

  intrinsic_proto::motion_planning::v1::TrajectoryPlanningResponse response;
  grpc::ClientContext ctx;
  INTR_RETURN_IF_ERROR(ToAbslStatus(
      motion_planner_service_->PlanTrajectory(&ctx, request, &response)));

  MotionPlannerClient::PlanTrajectoryResult result;
  result.trajectory = response.discretized();
  result.swept_volume.insert(result.swept_volume.begin(),
                             response.swept_volume().begin(),
                             response.swept_volume().end());
  result.lock_motion_id = response.has_lock_motion_id()
                              ? std::optional(response.lock_motion_id())
                              : std::nullopt;
  result.logging_id = response.logging_id();

  return result;
}

absl::StatusOr<std::vector<eigenmath::VectorXd>> MotionPlannerClient::ComputeIk(
    const world::KinematicObject& robot,
    const intrinsic_proto::motion_planning::CartesianMotionTarget&
        cartesian_target,
    const IkOptions& options) {
  // Convert CartesianMotionTarget to GeometricConstraint::PoseEquality.
  intrinsic_proto::motion_planning::v1::GeometricConstraint geometric_target;
  *geometric_target.mutable_cartesian_pose()->mutable_target_frame() =
      cartesian_target.frame();
  *geometric_target.mutable_cartesian_pose()->mutable_moving_frame() =
      cartesian_target.tool();
  if (cartesian_target.has_offset()) {
    *geometric_target.mutable_cartesian_pose()->mutable_target_frame_offset() =
        cartesian_target.offset();
  }

  return ComputeIk(robot, geometric_target, options);
}

absl::StatusOr<std::vector<eigenmath::VectorXd>> MotionPlannerClient::ComputeIk(
    const world::KinematicObject& robot,
    const intrinsic_proto::motion_planning::v1::GeometricConstraint&
        geometric_target,
    const IkOptions& options) {
  intrinsic_proto::motion_planning::v1::IkRequest request;
  request.set_world_id(world_id_);
  request.mutable_robot_reference()->mutable_object_id()->set_id(
      robot.Id().value());

  *request.mutable_target() = geometric_target;

  if (options.starting_joints.size() > 0) {
    VectorXdToRepeatedDouble(
        options.starting_joints,
        request.mutable_starting_joints()->mutable_joints());
  }

  if (options.max_num_solutions.has_value()) {
    request.set_max_num_solutions(*options.max_num_solutions);
  }

  if (options.collision_settings.has_value()) {
    *request.mutable_collision_settings() = *options.collision_settings;
  }

  request.set_ensure_same_branch(options.ensure_same_branch);

  request.set_prefer_same_branch(options.prefer_same_branch);

  intrinsic_proto::motion_planning::v1::IkResponse response;
  grpc::ClientContext ctx;
  INTR_RETURN_IF_ERROR(ToAbslStatus(
      motion_planner_service_->ComputeIk(&ctx, request, &response)));

  return ToVectorXds(response.solutions());
}

namespace {

// Common adaptor to handle different specifications of reference, target.
absl::StatusOr<Pose3d> ComputeFkInternal(
    const world::KinematicObject& robot,
    const eigenmath::VectorXd& joint_values,
    const intrinsic_proto::world::TransformNodeReference& reference,
    const intrinsic_proto::world::TransformNodeReference& target,
    const std::string& world_id,
    intrinsic_proto::motion_planning::v1::MotionPlannerService::StubInterface&
        motion_planner_service) {
  intrinsic_proto::motion_planning::v1::FkRequest request;
  request.set_world_id(world_id);
  request.mutable_robot_reference()->mutable_object_id()->set_id(
      robot.Id().value());
  VectorXdToRepeatedDouble(joint_values,
                           request.mutable_joints()->mutable_joints());

  *request.mutable_reference() = reference;
  *request.mutable_target() = target;

  intrinsic_proto::motion_planning::v1::FkResponse response;
  grpc::ClientContext ctx;
  INTR_RETURN_IF_ERROR(
      ToAbslStatus(motion_planner_service.ComputeFk(&ctx, request, &response)));

  return FromProto(response.reference_t_target());
}

}  // namespace

absl::StatusOr<Pose3d> MotionPlannerClient::ComputeFk(
    const world::KinematicObject& robot,
    const eigenmath::VectorXd& joint_values,
    const intrinsic_proto::world::TransformNodeReferenceByName& reference,
    const intrinsic_proto::world::TransformNodeReferenceByName& target) {
  intrinsic_proto::world::TransformNodeReference reference_proto;
  *reference_proto.mutable_by_name() = reference;
  intrinsic_proto::world::TransformNodeReference target_proto;
  *target_proto.mutable_by_name() = target;
  return ComputeFkInternal(robot, joint_values, reference_proto, target_proto,
                           world_id_, *motion_planner_service_);
}

absl::StatusOr<Pose3d> MotionPlannerClient::ComputeFk(
    const world::KinematicObject& robot,
    const eigenmath::VectorXd& joint_values,
    const world::TransformNode& reference, const world::TransformNode& target) {
  intrinsic_proto::world::TransformNodeReference reference_proto;
  reference_proto.set_id(reference.Id().value());
  intrinsic_proto::world::TransformNodeReference target_proto;
  target_proto.set_id(target.Id().value());
  return ComputeFkInternal(robot, joint_values, reference_proto, target_proto,
                           world_id_, *motion_planner_service_);
}

absl::StatusOr<intrinsic_proto::motion_planning::v1::CheckCollisionsResponse>
MotionPlannerClient::CheckCollisions(
    const world::KinematicObject& robot,
    const std::vector<eigenmath::VectorXd>& waypoints,
    const CheckCollisionsOptions& options) {
  intrinsic_proto::motion_planning::v1::CheckCollisionsRequest request;
  request.set_world_id(world_id_);
  request.mutable_robot_reference()->mutable_object_id()->set_id(
      robot.Id().value());
  ToJointVecs(waypoints, request.mutable_waypoint());

  if (options.collision_settings.has_value()) {
    *request.mutable_collision_settings() = *options.collision_settings;
  }

  intrinsic_proto::motion_planning::v1::CheckCollisionsResponse response;
  grpc::ClientContext ctx;
  INTR_RETURN_IF_ERROR(ToAbslStatus(
      motion_planner_service_->CheckCollisions(&ctx, request, &response)));
  return response;
}

absl::StatusOr<google::protobuf::Empty> MotionPlannerClient::ClearCache() {
  google::protobuf::Empty response;
  grpc::ClientContext ctx;
  INTR_RETURN_IF_ERROR(ToAbslStatus(motion_planner_service_->ClearCache(
      &ctx, google::protobuf::Empty(), &response)));
  return response;
}

}  // namespace motion_planning
}  // namespace intrinsic

File: intrinsic/motion_planning/motion_planner_client.py
# Copyright 2023 Intrinsic Innovation LLC

"""Defines the MotionPlannerClient class.

The MotionPlannerClient provides access to computations on top of the world.
This includes:
  * IK/FK
  * path planning
"""

import dataclasses
from typing import Any
from typing import Optional
from typing import Sequence

from google.protobuf import empty_pb2
# This import is required to use the *_grpc imports.
# pylint: disable=unused-import
import grpc

from intrinsic.assets.dependencies import utils as dep_utils
from intrinsic.assets.proto.v1 import resolved_dependency_pb2
from intrinsic.geometry.proto import transformed_geometry_storage_refs_pb2
from intrinsic.icon.proto import joint_space_pb2
from intrinsic.logging.proto import context_pb2
from intrinsic.math.python import data_types
from intrinsic.math.python import proto_conversion as math_proto_conversion
from intrinsic.motion_planning.proto import motion_target_pb2
from intrinsic.motion_planning.proto.v1 import compute_ik_pb2
from intrinsic.motion_planning.proto.v1 import geometric_constraints_pb2
from intrinsic.motion_planning.proto.v1 import motion_planner_config_pb2
from intrinsic.motion_planning.proto.v1 import motion_planner_service_pb2
from intrinsic.motion_planning.proto.v1 import motion_planner_service_pb2_grpc
from intrinsic.motion_planning.proto.v1 import motion_planning_pb2
from intrinsic.motion_planning.proto.v1 import motion_specification_pb2
from intrinsic.motion_planning.proto.v1 import robot_specification_pb2
from intrinsic.world.proto import collision_checker_config_pb2
from intrinsic.world.proto import collision_settings_pb2
from intrinsic.world.python import object_world_ids


def _repeated_vec_to_list_of_floats(
    vectors: list[joint_space_pb2.JointVec],
) -> list[list[float]]:
  return [list(vector.joints) for vector in vectors]


@dataclasses.dataclass
class CheckCollisionsOptions:
  """Options for Collision settings.

  Attributes:
    collision_settings: Settings for collision checking.
  """

  collision_settings: collision_settings_pb2.CollisionSettings | None = None


@dataclasses.dataclass
class IKOptions:
  """Options for IK.

  Attributes:
    max_num_solutions: The maximum number of solutions to be returned from IK
      computation. If not set (== 0), the underlying implementation has the
      freedom to choose. Negative values are invalid.
    starting_joints: The starting joint configuration to use. If not set, the
      current position of the robot in the world will be used.
    collision_settings: Collision Settings. If left empty, no collision checking
      is done.
    ensure_same_branch: Flag to choose IK solution is on the same kinematic
      branch as the starting joints of the robot.
    prefer_same_branch:  Flag that will prefer solutions on the same kinematic
      branch over those close to the starting_joints configuration.
    disable_error_on_collisions: If True, does not return an error when a
      starting state is in collision or when the robot cannot escape a
      collision.
    collision_checker_config: Configuration override for the collision checker.
    context: The logging context for the skill sending the request.
    disable_logging: If True, disables logging of the ComputeIk request data.
    caller_id: Optional caller ID to correlate the IKRequest logs with its caller.
  """

  max_num_solutions: int = 0
  starting_joints: list[float] | None = None
  collision_settings: collision_settings_pb2.CollisionSettings | None = None
  ensure_same_branch: bool = False
  prefer_same_branch: bool = False
  disable_error_on_collisions: bool = False
  collision_checker_config: (
      collision_checker_config_pb2.CollisionCheckerConfig | None
  ) = None
  context: context_pb2.Context | None = None
  disable_logging: bool = False
  caller_id: str | None = None


@dataclasses.dataclass
class MotionPlanningOptions:
  """Options for Motion Planning.

  Attributes:
    path_planning_time_out: Timeout for path planning algorithms in seconds.
    path_planning_step_size: Maximum step size deployed during path planning. If
      not specified, a default step size is used.
    lock_motion_configuration: Optional configuration for saving or loading a
      motion.
    skip_fuzzy_cache_check: If true, the cache will not check for fuzzy matches.
    compute_swept_volume: Optionally generate and return the swept volume for
      the computed path.
    shortcutting_combine_collinear_segments: If true, the joint_shortcutter will
      combine multiple collinear segments into a single segment. This speeds up
      planning time, but can result in longer paths and trajectories.
    collision_checker_config: Configuration override for the collision checker.
  """

  path_planning_time_out: int = 30
  path_planning_step_size: float | None = None
  lock_motion_configuration: (
      motion_planner_config_pb2.LockMotionConfiguration | None
  ) = None
  skip_fuzzy_cache_check: bool = False
  compute_swept_volume: bool = False
  shortcutting_combine_collinear_segments: bool = False
  collision_checker_config: (
      collision_checker_config_pb2.CollisionCheckerConfig | None
  ) = None


@dataclasses.dataclass
class PlanTrajectoryResult:
  """Wrapped result from calling plan_trajectory.

  Attributes:
    trajectory: The computed joint trajectory containing position, velocity,
      and acceleration.
    swept_volume: An optional list of shapes that correspond to the swept
      volume of the trajectory.
    lock_motion_id: If the motion is locked based on the request, this is the ID
      of the motion for loading later. Note this field is not set if this
      response is from loading a locked motion.
  """

  trajectory: joint_space_pb2.JointTrajectoryPVA
  swept_volume: list[
      transformed_geometry_storage_refs_pb2.TransformedGeometryStorageRefs
  ]
  lock_motion_id: Optional[str] = None
  logging_id: str = ""

@dataclasses.dataclass
class PlanPathResult:
  """Wrapped result from calling plan_path."""

  path: motion_planning_pb2.Path
  swept_volume: list[
      transformed_geometry_storage_refs_pb2.TransformedGeometryStorageRefs
  ]
  logging_id: str = ""

@dataclasses.dataclass
class ComputeIkResult:
  """Wrapped result from calling compute_ik.

  Attributes:
    solutions: The computed IK solutions (joint configurations).
    ik_debug_information: Detailed internal diagnostics and metrics from the IK
      solver. NOTE: Unintuitively, the `ik_solutions` list within this debug
      structure contains all candidate configurations generated and evaluated
      during the search, including those that were ultimately REJECTED (e.g.,
      due to collisions, joint limit violations, or constraint failures).
    logging_id: Unique identifier generated by the service for tracing log
      entries.
  """

  solutions: list[list[float]]
  ik_debug_information: Optional[compute_ik_pb2.ComputeIkDebugInformation] = (
      None
  )
  logging_id: str = ""


class MotionPlannerClientBase:
  """The base client for the MotionPlannerService.

  The class includes the externalized functionality of the MotionPlannerClient.
  """

  def __init__(
      self,
      world_id: str,
      stub: motion_planner_service_pb2_grpc.MotionPlannerServiceStub,
  ):
    self._world_id: str = world_id
    self._stub: motion_planner_service_pb2_grpc.MotionPlannerServiceStub = stub

  def clear_cache(self) -> empty_pb2.Empty:
    """Calls the ClearCache rpc."""
    return self._stub.ClearCache(empty_pb2.Empty())


class MotionPlannerClient(MotionPlannerClientBase):
  """Helper class for calling the rpcs in the MotionPlannerService.

  Provides additional computations on top of the world.
    * IK/FK
    * Path planning
  """

  def plan_trajectory(
      self,
      robot_specification: robot_specification_pb2.RobotSpecification,
      motion_specification: motion_specification_pb2.MotionSpecification,
      options: MotionPlanningOptions = MotionPlanningOptions(),
      caller_id: str = "Anonymous",
  ) -> PlanTrajectoryResult:
    """Plan trajectory for a given motion planning problem and robot.

    This method calls the Plan trajectory rpc.

    Args:
      robot_specification: Robot specification
      motion_specification: Motion specification, see MotionSpecification proto.
      options: Motion planning options that allows the path planning timeout to
        be set.
      caller_id: The id used for logging the request in the motion planner
        service.

    Returns:
      Discretized trajectory
    """
    request = motion_planner_service_pb2.MotionPlanningRequest(
        world_id=self._world_id,
        robot_specification=robot_specification,
        motion_specification=motion_specification,
        caller_id=caller_id,
    )
    request.motion_planner_config.timeout_sec.seconds = (
        options.path_planning_time_out
    )
    request.compute_swept_volume = options.compute_swept_volume

    if options.path_planning_step_size is not None:
      request.motion_planner_config.path_planning_step_size = (
          options.path_planning_step_size
      )
    if options.lock_motion_configuration:
      request.motion_planner_config.lock_motion_configuration.CopyFrom(
          options.lock_motion_configuration
      )
    if options.skip_fuzzy_cache_check:
      request.motion_planner_config.skip_fuzzy_cache_check = (
          options.skip_fuzzy_cache_check
      )
    if options.shortcutting_combine_collinear_segments:
      request.motion_planner_config.shortcutting_combine_collinear_segments = (
          options.shortcutting_combine_collinear_segments
      )
    if options.collision_checker_config is not None:
      request.motion_planner_config.collision_checker_config.CopyFrom(
          options.collision_checker_config
      )
    response = self._stub.PlanTrajectory(request)
    swept_volume = list(response.swept_volume)
    lock_motion_id = (
        response.lock_motion_id if response.HasField("lock_motion_id") else None
    )
    return PlanTrajectoryResult(
        trajectory=response.discretized,
        swept_volume=swept_volume,
        lock_motion_id=lock_motion_id,
        logging_id=response.logging_id,
    )

  def plan_path(
      self,
      robot_specification: robot_specification_pb2.RobotSpecification,
      motion_specification: motion_specification_pb2.MotionSpecification,
      options: MotionPlanningOptions = MotionPlanningOptions(),
      caller_id: str = "Anonymous",
  ) -> PlanPathResult:
    """Plan path for a given motion planning problem and robot.

    This method calls the Plan Path rpc.

    Args:
      robot_specification: Robot specification
      motion_specification: Motion specification, see MotionSpecification proto.
      options: Motion planning options that allows the path planning timeout to
        be set.
      caller_id: The id used for logging the request in the motion planner
        service.

    Returns:
      Planned path consisting of waypoints.
    """
    request = motion_planner_service_pb2.MotionPlanningRequest(
        world_id=self._world_id,
        robot_specification=robot_specification,
        motion_specification=motion_specification,
        caller_id=caller_id,
    )
    request.motion_planner_config.timeout_sec.seconds = (
        options.path_planning_time_out
    )
    request.compute_swept_volume = options.compute_swept_volume

    if options.path_planning_step_size is not None:
      request.motion_planner_config.path_planning_step_size = (
          options.path_planning_step_size
      )
    if options.lock_motion_configuration:
      request.motion_planner_config.lock_motion_configuration.CopyFrom(
          options.lock_motion_configuration
      )
    if options.skip_fuzzy_cache_check:
      request.motion_planner_config.skip_fuzzy_cache_check = (
          options.skip_fuzzy_cache_check
      )
    if options.shortcutting_combine_collinear_segments:
      request.motion_planner_config.shortcutting_combine_collinear_segments = (
          options.shortcutting_combine_collinear_segments
      )
    if options.collision_checker_config is not None:
      request.motion_planner_config.collision_checker_config.CopyFrom(
          options.collision_checker_config
      )
    response = self._stub.PlanPath(request)
    swept_volume = list(response.swept_volume)
    return PlanPathResult(
        path=response.path,
        swept_volume=swept_volume,
        logging_id=response.logging_id,
    )

  def compute_ik(
      self,
      robot_name: object_world_ids.WorldObjectName,
      target: (
          motion_target_pb2.CartesianMotionTarget
          | geometric_constraints_pb2.GeometricConstraint
      ),
      starting_joints: Optional[list[float]] = None,
      options: Optional[IKOptions] = None,
  ) -> ComputeIkResult:
    """Calls the ComputeIk rpc, doing argument conversion as necessary.

    Args:
      robot_name: Name of robot, must map to a kinematic object
      target: a target pose to compute ik for
      starting_joints: used as seed for ik, optional
      options: See IKOptions.

    Returns:
      A list of IK solutions
    """
    request = motion_planner_service_pb2.IkRequest(world_id=self._world_id)
    request.robot_reference.object_id.by_name.object_name = robot_name

    # Convert CartesianMotionTarget to GeometricConstraint::PoseEquality.
    if isinstance(target, motion_target_pb2.CartesianMotionTarget):
      request.target.cartesian_pose.moving_frame.CopyFrom(target.tool)
      request.target.cartesian_pose.target_frame.CopyFrom(target.frame)
      if target.HasField("offset"):
        request.target.cartesian_pose.target_frame_offset.CopyFrom(
            target.offset
        )
    else:
      request.target.CopyFrom(target)

    if starting_joints is not None:
      request.starting_joints.joints.extend(starting_joints)
    if options is not None:
      if options.collision_settings:
        request.collision_settings.CopyFrom(options.collision_settings)
      if options.ensure_same_branch:
        request.ensure_same_branch = options.ensure_same_branch
      if options.prefer_same_branch:
        request.prefer_same_branch = options.prefer_same_branch
      if options.max_num_solutions:
        request.max_num_solutions = options.max_num_solutions
      if options.disable_error_on_collisions:
        request.disable_error_on_collisions = (
            options.disable_error_on_collisions
        )
      if options.collision_checker_config:
        request.collision_checker_config.CopyFrom(
            options.collision_checker_config
        )
      if options.context:
        request.context.CopyFrom(options.context)
      if options.disable_logging:
        request.disable_logging = options.disable_logging
      if options.caller_id:
        request.caller_id = options.caller_id

    # Make the rpc.
    response = self._stub.ComputeIk(request)
    return ComputeIkResult(
        solutions=_repeated_vec_to_list_of_floats(response.solutions),
        ik_debug_information=response.ik_debug_information,
        logging_id=response.logging_id,
    )

  def compute_fk(
      self,
      robot_name: object_world_ids.WorldObjectName,
      joints: list[float],
      reference: object_world_ids.WorldObjectName,
      target: object_world_ids.WorldObjectName,
      reference_frame: Optional[object_world_ids.FrameName] = None,
      target_frame: Optional[object_world_ids.FrameName] = None,
  ) -> data_types.Pose3:
    """Calls the ComputeIk rpc, doing argument conversion as necessary.

    If reference_frame is not specified, the WorldObject 'reference' will be
    used directly. The same applies for target_frame/target.

    Args:
      robot_name: Name of robot, must map to a kinematic object
      joints: the joint configuration values to compute fk for
      reference: object name for the reference of the returned pose
      target: object name for the target of the returned pose
      reference_frame: name that specifies a frame under 'reference', optional
      target_frame: name that specifies a frame under 'target', optional

    Returns:
      The reference_t_target pose, i.e. the pose of 'target' in the frame of
      'reference'
    """
    request = motion_planner_service_pb2.FkRequest(world_id=self._world_id)
    request.robot_reference.object_id.by_name.object_name = robot_name
    request.joints.joints.extend(joints)
    if reference_frame is not None:
      # If frame is not None, then we'll assume the user is specifying a frame,
      # and otherwise the user is specifying an object.
      request.reference.by_name.frame.object_name = reference
      request.reference.by_name.frame.frame_name = reference_frame
    else:
      request.reference.by_name.object.object_name = reference
    if target_frame is not None:
      # If frame is not None, then we'll assume the user is specifying a frame,
      # and otherwise the user is specifying an object.
      request.target.by_name.frame.object_name = target
      request.target.by_name.frame.frame_name = target_frame
    else:
      request.target.by_name.object.object_name = target

    # Make the rpc.
    response = self._stub.ComputeFk(request)
    return math_proto_conversion.pose_from_proto(response.reference_t_target)

  def check_collisions(
      self,
      robot_name: object_world_ids.WorldObjectName,
      waypoints: list[joint_space_pb2.JointVec],
      options: CheckCollisionsOptions,
  ) -> motion_planner_service_pb2.CheckCollisionsResponse:
    """Calls the CheckCollisions rpc.

    Args:
      robot_name: Name of robot, must map to a kinematic object.
      waypoints: The waypoints define the path for which we check the collision.
        We also check the linear interpolation between the waypoints.
      options: Check collision options

    Returns:
      CheckCollisionResponse. See flag `has_collision` in response
      which indicates a collision.
    """

    request = motion_planner_service_pb2.CheckCollisionsRequest(
        world_id=self._world_id,
        robot_reference=robot_specification_pb2.RobotReference(),
        waypoint=waypoints,
        collision_settings=options.collision_settings,
    )
    request.robot_reference.object_id.by_name.object_name = robot_name
    response = self._stub.CheckCollisions(request)
    return response


def get_motion_planner_service_asset_client(
    world_id: str,
    mps_dependency: resolved_dependency_pb2.ResolvedDependency,
    grpc_options: Optional[Sequence[tuple[str, Any]]] = None,
) -> Optional[MotionPlannerClient]:
  """Returns a MotionPlannerClient connected to the MotionPlannerService Asset.

  Args:
    world_id: The world ID.
    mps_dependency: The resolved dependency.
    grpc_options: Optional gRPC options.

  Returns:
    A MotionPlannerClient or None if mps_dependency has no interfaces.
  """
  if not mps_dependency.interfaces:
    return None

  # Dynamically find the MotionPlannerService interface key.
  iface = None
  for key in mps_dependency.interfaces:
    if "MotionPlannerService" in key:
      iface = key
      break

  if iface is None:
    raise ValueError(
        "Could not find MotionPlannerService interface in resolved dependency. "
        f"Available interfaces: {list(mps_dependency.interfaces.keys())}"
    )

  channel = dep_utils.connect(
      mps_dependency,
      iface,
      grpc_options=grpc_options,
  )
  stub = motion_planner_service_pb2_grpc.MotionPlannerServiceStub(channel)
  return MotionPlannerClient(world_id, stub)
